src/Microsoft.OData.Core/ODataBatchWriter.cs

```diff
@@ -422,473 +388,307 @@ void IODataBatchOperationListener.BatchOperationContentStreamDisposed()
 
-            return this.CurrentOperationResponseMessage;
-        }
+        /// <summary>
+        /// Writes the start boundary for an operation. This is either the batch or the changeset boundary.
+        /// </summary>
+        protected abstract void WriteStartBoundaryForOperation();
 
         /// <summary>
-        /// Writes all the pending headers and prepares the writer to write a content of the operation.
+        /// Determines whether a given writer state is considered an error state.
         /// </summary>
-        private void StartBatchOperationContent()
+        /// <param name="state">The writer state to check.</param>
+        /// <returns>True if the writer state is an error state; otherwise false.</returns>
+        private static bool IsErrorState(BatchWriterState state)
         {
-            Debug.Assert(this.CurrentOperationMessage != null, "Expected non-null operation message!");
-            Debug.Assert(this.rawOutputContext.TextWriter != null, "Must have a batch writer!");
-
-            // write the pending headers (if any)
-            this.WritePendingMessageData(false);
-
-            // flush the text writer to make sure all buffers of the text writer
-            // are flushed to the underlying async stream
-            this.rawOutputContext.TextWriter.Flush();
+            return state == BatchWriterState.Error;
         }
 
         /// <summary>
-        /// Disposes the batch writer and set the 'OperationStreamRequested' batch writer state;
-        /// called after the flush operation(s) have completed.
+        /// Verifies that calling WriteStartBatch is valid.
         /// </summary>
-        private void DisposeBatchWriterAndSetContentStreamRequestedState()
+        /// <param name="synchronousCall">true if the call is to be synchronous; false otherwise.</param>
+        private void VerifyCanWriteStartBatch(bool synchronousCall)
         {
-            this.rawOutputContext.CloseWriter();
-
-            this.SetState(BatchWriterState.OperationStreamRequested);
+            this.ValidateWriterReady();
+            this.VerifyCallAllowed(synchronousCall);
         }
 
         /// <summary>
-        /// Remember a non-null Content-ID header for change set request operations.
-        /// If a non-null content ID header is specified for a change set request operation, record it in the URL resolver.
+        /// Starts a new batch - implementation of the actual functionality.
         /// </summary>
-        /// <param name="contentId">The Content-ID header value read from the message.</param>
-        /// <remarks>
-        /// Note that the content ID of this operation will only
-        /// become visible once this operation has been written
-        /// and OperationCompleted has been called on the URL resolver.
-        /// </remarks>
-        private void RememberContentIdHeader(string contentId)
+        private void WriteStartBatchImplementation()

```

Comment:
>WriteStartBatchImplementation [](start = 21, length = 29)

This should be protected abstract, so derived class can overwrite to implement some custom logic in writing start of batch.

Reply:
Actually, this method should only set the state, generic enough so that we don't expect to be overridden, therefore private in base class is OK.

---
In reply to: [140933843](https://github.com/OData/odata.net/pull/960#discussion_r140933843) [](ancestors = 140933843)

Reply:
In the multipart/mixed there is no reason to do anything at the start of the batch other than set the state. In other formats (including json) there is work that the derived class may want to do in order to start the batch (like write the start object and "requests" property in JSON).

---
In reply to: [142561309](https://github.com/OData/odata.net/pull/960#discussion_r142561309) [](ancestors = 142561309,140933843)